### PR TITLE
[RISCV] Custom promote s32 G_UDIV/UREM/SDIV on RV64. Promote SREM using G_SEXT.

### PR DESCRIPTION
**Pull request by topperc**
We don't add a custom node for SREM as we can detect it with (srem (sexti32), (sexti32))

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/alu_m-rv64.mir [hunk 1]
```diff
@@ -47,39 +47,9 @@ body:             |
...
-    %1:gprb(s32) = G_TRUNC %0(s64)
-    %2:gprb(s64) = COPY $x11
-    %3:gprb(s32) = G_TRUNC %2(s64)
-    %4:gprb(s32) = G_SDIV %1, %3
-    %5:gprb(s64) = G_ANYEXT %4(s32)
-    $x10 = COPY %5(s64)
-    PseudoRET implicit $x10
-
-...
----
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/alu_m-rv64.mir:59]: I removed the srem i32 test here but it's still covered by an IR test.

File: llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/alu_m-rv64.mir [hunk 2]
```diff
@@ -124,16 +91,13 @@ body:             |
...
     ; RV64I-NEXT: PseudoRET implicit $x10
     %0:gprb(s64) = COPY $x10
-    %1:gprb(s32) = G_TRUNC %0(s64)
-    %2:gprb(s64) = COPY $x11
-    %3:gprb(s32) = G_TRUNC %2(s64)
-    %4:gprb(s32) = G_UREM %1, %3
-    %5:gprb(s64) = G_ANYEXT %4(s32)
-    $x10 = COPY %5(s64)
+    %1:gprb(s64) = COPY $x11
+    %2:gprb(s64) = G_UREM %0, %1
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/alu_m-rv64.mir:99]: There's a mistake here

**Comment by mshockwave** [comment]
> as we can detect it with (srem (sexti32), (sexti32))

as a result of `clampScalar(0, sXLen, sDoubleXLen)`?

File: llvm/lib/Target/RISCV/RISCVInstrGISel.td [hunk 3]
```diff
@@ -17,6 +17,30 @@ class RISCVGenericInstruction : GenericInstruction {
...
+}
+def : GINodeEquiv<G_DIVUW, riscv_divuw>;
+
+// Pseudo equivalent to a RISCVISD::REMUW.
+def G_REMUW : RISCVGenericInstruction {
+  let OutOperandList = (outs type0:$dst);
+  let InOperandList = (ins type0:$src1, type0:$src2);
+  let hasSideEffects = false;
+}
+def : GINodeEquiv<G_REMUW, riscv_remuw>;
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/RISCVInstrGISel.td:42]: we probably can factor out RISCVGenericInstruction of this kind into a separate TableGen class in the future

**Review by mshockwave — APPROVED** [review]
LGTM

**Comment by topperc** [comment]
> > as we can detect it with (srem (sexti32), (sexti32))
> 
> as a result of `clampScalar(0, sXLen, sDoubleXLen)`?

That's how the type is widened, but the comment was about the math.

The important thing is that the behavior of (srem (sexti32 X), (sexti32 Y)) is exactly equivalent to REMW for all values of X and Y. From computeNumSignBits for srem.

```
    // The sign bit is the LHS's sign bit, except when the result of the         
    // remainder is zero. The magnitude of the result should be less than or     
    // equal to the magnitude of the LHS. Therefore, the result should have      
    // at least as many sign bits as the left hand side.
```

The result of (sexti32 X) has at least 33 sign bits so the result of (srem (sexti32 X), (sexti32 Y)) has at least 33 sign bits. REMW produces a value with at least 33 sign bits.

This doesn't work for (sdiv (sexti32 X), (sexti32 Y)). 0xffffffff80000000 / 0xfffffffffffffff is 0x0000000080000000, but DIVW produces 0xffffffff80000000 for that case. We use a custom opcode to remember that we started with i32 where 0x80000000 / 0xffffffff is immediate UB so that case doesn't matter.

(urem (zexti32 X), (zexti32 Y)) is not equivalent to REMUW because 0x0000000080000000 % 0x00000000ffffffff is 0x0000000080000000, but REMUW will produce 0xffffffff80000000. Since we started with an i32 type, the upper bits don't matter at the time of promotion so we use a custom opcode to remember that. After promotion, later optimizations may become dependent on the G_REMUW producing a sign extended value. Though I haven't addded that to computeNumSignBits yet for GISel.

(udiv (zexti32 X), (zexti32 Y)) is not equivalent to DIVUW because 0x0000000080000000 % 0x0000000000000001 is 0x0000000080000000, but DIVUW will return 0xffffffff80000000. Again at promotion time the upper don't matter so we use a custom opcode.

**Pull request merged**